cpp/dolfinx/fem/Form.h

```diff
@@ -39,17 +41,109 @@ enum class IntegralType : std::int8_t
   vertex = 3          ///< Vertex
 };
 
+namespace impl
+{
+/// @brief Compute the list of entity indices in `mesh` for the ith
+/// integral (kernel) of a given type (i.e. cell, exterior facet, or
+/// interior facet).
+inline std::vector<std::int32_t>
+compute_domain(IntegralType type, const mesh::Topology& topology, int codim,
+               std::span<const std::int32_t> entities,
+               std::span<const std::int32_t> entity_map)
+{
+  assert(codim >= 0);
+  const int tdim = topology.dim();
+
+  std::vector<std::int32_t> mapped_entities;
+  mapped_entities.reserve(entities.size());
+  switch (type)
+  {
+  case IntegralType::cell:
+  {
+    std::ranges::transform(entities, std::back_inserter(mapped_entities),
+                           [&entity_map](auto e) { return entity_map[e]; });
+    break;
+  }
+  case IntegralType::exterior_facet:
+  {
+    if (codim == 0)
+    {
+      for (std::size_t i = 0; i < entities.size(); i += 2)
+      {
+        // Add cell and the local facet index
+        mapped_entities.insert(mapped_entities.end(),
+                               {entity_map[entities[i]], entities[i + 1]});
+      }
+    }
+    else if (codim == 1)
+    {
+      // In this case, the entity maps take facets in (`_mesh`) to cells
+      // in `mesh`, so we need to get the facet number from the (cell,
+      // local_facet pair) first.
+      auto c_to_f = topology.connectivity(tdim, tdim - 1);
+      assert(c_to_f);
+      for (std::size_t i = 0; i < entities.size(); i += 2)
+      {
+        // Get the facet index, and add cell and the local facet index
+        std::int32_t facet = c_to_f->links(entities[i])[entities[i + 1]];
+        mapped_entities.insert(mapped_entities.end(),
+                               {entity_map[facet], entities[i + 1]});
```

Comment:
Is this correct?
Say we have an exterior facet integral `(cell_0, facet_index_0)` which should be mapped onto a submesh of codim-1. Then the integration entity on the "submesh" is a cell integral for that mesh, and thus the "local facet index" (entities[i+1])" doesn;t really make sense to store. @jpdean am I missing some logic here?

Reply:
Code is unchanged from `main`, https://github.com/FEniCS/dolfinx/blob/ead49d0fdd57d2273162ede3484505fa77981001/cpp/dolfinx/fem/Form.h#L511.

Reply:
I am not stating that you have introduced a bug. What I am trying to say is that I don't think we need to add entities[i+1]  for codim 1, and that we could send in 0. I'll let @jpdean comment if he agrees.

Reply:
Yes, I agree. Though it's probably not a major issue for now